# SQL assignment on kaggle dataset  "Sample_Superstore.csv"

In [1]:
import sqlite3
import pandas as pd

conn = sqlite3.connect("celebal_internship.db")

In [4]:
df = pd.read_csv("Sample_Superstore.csv", encoding="latin1")

In [5]:
df.head()

,Row ID,Order ID,Order Date,Ship Date,Ship Mode,Customer ID,Customer Name,Segment,Country,City,...,Postal Code,Region,Product ID,Category,Sub-Category,Product Name,Sales,Quantity,Discount,Profit
0,1,CA-2016-152156,11/8/2016,11/11/2016,Second Class,CG-12520,Claire Gute,Consumer,United States,Henderson,...,42420,South,FUR-BO-10001798,Furniture,Bookcases,Bush Somerset Collection Bookcase,261.9600,2,0.00,41.9136
1,2,CA-2016-152156,11/8/2016,11/11/2016,Second Class,CG-12520,Claire Gute,Consumer,United States,Henderson,...,42420,South,FUR-CH-10000454,Furniture,Chairs,"Hon Deluxe Fabric Upholstered Stacking Chairs,...",731.9400,3,0.00,219.5820
2,3,CA-2016-138688,6/12/2016,6/16/2016,Second Class,DV-13045,Darrin Van Huff,Corporate,United States,Los Angeles,...,90036,West,OFF-LA-10000240,Office Supplies,Labels,Self-Adhesive Address Labels for Typewriters b...,14.6200,2,0.00,6.8714
3,4,US-2015-108966,10/11/2015,10/18/2015,Standard Class,SO-20335,Sean O'Donnell,Consumer,United States,Fort Lauderdale,...,33311,South,FUR-TA-10000577,Furniture,Tables,Bretford CR4500 Series Slim Rectangular Table,957.5775,5,0.45,-383.0310
4,5,US-2015-108966,10/11/2015,10/18/2015,Standard Class,SO-20335,Sean O'Donnell,Consumer,United States,Fort Lauderdale,...,33311,South,OFF-ST-10000760,Office Supplies,Storage,Eldon Fold 'N Roll Cart System,22.3680,2,0.20,2.5164


## 1.Load dataset into a SQL database

In [6]:
df.to_sql(
    "sales_data",
    conn,
    if_exists="replace",
    index=False
)

9994

## 2.Explore table (schema, sample data)

### Total Records

In [8]:
query = """
SELECT COUNT(*) AS total_rows
FROM sales_data
"""

pd.read_sql(query, conn)

,total_rows
0,9994


The dataset contains 9994 records which are used for further queries.

### First 10 records

In [9]:
query = """
SELECT *
FROM sales_data
LIMIT 10
"""

pd.read_sql(query, conn)

,Row ID,Order ID,Order Date,Ship Date,Ship Mode,Customer ID,Customer Name,Segment,Country,City,...,Postal Code,Region,Product ID,Category,Sub-Category,Product Name,Sales,Quantity,Discount,Profit
0,1,CA-2016-152156,11/8/2016,11/11/2016,Second Class,CG-12520,Claire Gute,Consumer,United States,Henderson,...,42420,South,FUR-BO-10001798,Furniture,Bookcases,Bush Somerset Collection Bookcase,261.9600,2,0.00,41.9136
1,2,CA-2016-152156,11/8/2016,11/11/2016,Second Class,CG-12520,Claire Gute,Consumer,United States,Henderson,...,42420,South,FUR-CH-10000454,Furniture,Chairs,"Hon Deluxe Fabric Upholstered Stacking Chairs,...",731.9400,3,0.00,219.5820
2,3,CA-2016-138688,6/12/2016,6/16/2016,Second Class,DV-13045,Darrin Van Huff,Corporate,United States,Los Angeles,...,90036,West,OFF-LA-10000240,Office Supplies,Labels,Self-Adhesive Address Labels for Typewriters b...,14.6200,2,0.00,6.8714
3,4,US-2015-108966,10/11/2015,10/18/2015,Standard Class,SO-20335,Sean O'Donnell,Consumer,United States,Fort Lauderdale,...,33311,South,FUR-TA-10000577,Furniture,Tables,Bretford CR4500 Series Slim Rectangular Table,957.5775,5,0.45,-383.0310
4,5,US-2015-108966,10/11/2015,10/18/2015,Standard Class,SO-20335,Sean O'Donnell,Consumer,United States,Fort Lauderdale,...,33311,South,OFF-ST-10000760,Office Supplies,Storage,Eldon Fold 'N Roll Cart System,22.3680,2,0.20,2.5164
5,6,CA-2014-115812,6/9/2014,6/14/2014,Standard Class,BH-11710,Brosina Hoffman,Consumer,United States,Los Angeles,...,90032,West,FUR-FU-10001487,Furniture,Furnishings,Eldon Expressions Wood and Plastic Desk Access...,48.8600,7,0.00,14.1694
6,7,CA-2014-115812,6/9/2014,6/14/2014,Standard Class,BH-11710,Brosina Hoffman,Consumer,United States,Los Angeles,...,90032,West,OFF-AR-10002833,Office Supplies,Art,Newell 322,7.2800,4,0.00,1.9656
7,8,CA-2014-115812,6/9/2014,6/14/2014,Standard Class,BH-11710,Brosina Hoffman,Consumer,United States,Los Angeles,...,90032,West,TEC-PH-10002275,Technology,Phones,Mitel 5320 IP Phone VoIP phone,907.1520,6,0.20,90.7152
8,9,CA-2014-115812,6/9/2014,6/14/2014,Standard Class,BH-11710,Brosina Hoffman,Consumer,United States,Los Angeles,...,90032,West,OFF-BI-10003910,Office Supplies,Binders,DXL Angle-View Binders with Locking Rings by S...,18.5040,3,0.20,5.7825
9,10,CA-2014-115812,6/9/2014,6/14/2014,Standard Class,BH-11710,Brosina Hoffman,Consumer,United States,Los Angeles,...,90032,West,OFF-AP-10002892,Office Supplies,Appliances,Belkin F5C206VTEL 6 Outlet Surge,114.9000,5,0.00,34.4700


Displays the first 10 records to understand the structure and contents of the table.

### Table Structure and data types

In [11]:
pd.read_sql(
    "PRAGMA table_info(sales_data)",
    conn
)

,cid,name,type,notnull,dflt_value,pk
0,0,Row ID,INTEGER,0,None,0
1,1,Order ID,TEXT,0,None,0
2,2,Order Date,TEXT,0,None,0
3,3,Ship Date,TEXT,0,None,0
4,4,Ship Mode,TEXT,0,None,0
5,5,Customer ID,TEXT,0,None,0
6,6,Customer Name,TEXT,0,None,0
7,7,Segment,TEXT,0,None,0
8,8,Country,TEXT,0,None,0
9,9,City,TEXT,0,None,0


Reviewed the column names and datatypes before performing analysis.

## 3.Apply WHERE filters (region, category, date, sales)

### Filtering data of West region

In [17]:
query = """
SELECT *
FROM sales_data
WHERE Region = 'West'
"""

pd.read_sql(query, conn)

,Row ID,Order ID,Order Date,Ship Date,Ship Mode,Customer ID,Customer Name,Segment,Country,City,...,Postal Code,Region,Product ID,Category,Sub-Category,Product Name,Sales,Quantity,Discount,Profit
0,3,CA-2016-138688,6/12/2016,6/16/2016,Second Class,DV-13045,Darrin Van Huff,Corporate,United States,Los Angeles,...,90036,West,OFF-LA-10000240,Office Supplies,Labels,Self-Adhesive Address Labels for Typewriters b...,14.620,2,0.0,6.8714
1,6,CA-2014-115812,6/9/2014,6/14/2014,Standard Class,BH-11710,Brosina Hoffman,Consumer,United States,Los Angeles,...,90032,West,FUR-FU-10001487,Furniture,Furnishings,Eldon Expressions Wood and Plastic Desk Access...,48.860,7,0.0,14.1694
2,7,CA-2014-115812,6/9/2014,6/14/2014,Standard Class,BH-11710,Brosina Hoffman,Consumer,United States,Los Angeles,...,90032,West,OFF-AR-10002833,Office Supplies,Art,Newell 322,7.280,4,0.0,1.9656
3,8,CA-2014-115812,6/9/2014,6/14/2014,Standard Class,BH-11710,Brosina Hoffman,Consumer,United States,Los Angeles,...,90032,West,TEC-PH-10002275,Technology,Phones,Mitel 5320 IP Phone VoIP phone,907.152,6,0.2,90.7152
4,9,CA-2014-115812,6/9/2014,6/14/2014,Standard Class,BH-11710,Brosina Hoffman,Consumer,United States,Los Angeles,...,90032,West,OFF-BI-10003910,Office Supplies,Binders,DXL Angle-View Binders with Locking Rings by S...,18.504,3,0.2,5.7825
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
3198,9987,CA-2016-125794,9/29/2016,10/3/2016,Standard Class,ML-17410,Maris LaWare,Consumer,United States,Los Angeles,...,90008,West,TEC-AC-10003399,Technology,Accessories,Memorex Mini Travel Drive 64 GB USB 2.0 Flash ...,36.240,1,0.0,15.2208
3199,9991,CA-2017-121258,2/26/2017,3/3/2017,Standard Class,DB-13060,Dave Brooks,Consumer,United States,Costa Mesa,...,92627,West,FUR-FU-10000747,Furniture,Furnishings,Tenex B1-RE Series Chair Mats for Low Pile Car...,91.960,2,0.0,15.6332
3200,9992,CA-2017-121258,2/26/2017,3/3/2017,Standard Class,DB-13060,Dave Brooks,Consumer,United States,Costa Mesa,...,92627,West,TEC-PH-10003645,Technology,Phones,Aastra 57i VoIP phone,258.576,2,0.2,19.3932
3201,9993,CA-2017-121258,2/26/2017,3/3/2017,Standard Class,DB-13060,Dave Brooks,Consumer,United States,Costa Mesa,...,92627,West,OFF-PA-10004041,Office Supplies,Paper,"It's Hot Message Books with Stickers, 2 3/4"" x 5""",29.600,4,0.0,13.3200


The above  query shows all the details of the West region.

### Filtering data with sales>1000

In [20]:
query = """
SELECT
"Order ID" , "Customer Name" , Category, "Product Name" , Sales , Profit
FROM sales_data
WHERE Sales > 1000
ORDER BY Sales DESC;
"""

pd.read_sql(query, conn)

,Order ID,Customer Name,Category,Product Name,Sales,Profit
0,CA-2014-145317,Sean Miller,Technology,Cisco TelePresence System EX90 Videoconferenci...,22638.480,-1811.0784
1,CA-2016-118689,Tamara Chand,Technology,Canon imageCLASS 2200 Advanced Copier,17499.950,8399.9760
2,CA-2017-140151,Raymond Buch,Technology,Canon imageCLASS 2200 Advanced Copier,13999.960,6719.9808
3,CA-2017-127180,Tom Ashbrook,Technology,Canon imageCLASS 2200 Advanced Copier,11199.968,3919.9888
4,CA-2017-166709,Hunter Lopez,Technology,Canon imageCLASS 2200 Advanced Copier,10499.970,5039.9856
...,...,...,...,...,...,...
463,CA-2017-147956,Alan Hwang,Office Supplies,Office Impressions Heavy Duty Welded Shelving ...,1003.620,0.0000
464,CA-2017-163006,Gary Hansen,Technology,Samsung Galaxy S4,1001.584,125.1980
465,CA-2016-140249,Shaun Weien,Technology,Samsung Galaxy S4,1001.584,125.1980
466,CA-2017-138380,Yana Sorensen,Office Supplies,Letter Size Cart,1000.020,290.0058


This query identifies high-value orders with sales greater than 1000. It helps highlight premium products and customers contributing significantly to revenue.

## 4.Use GROUP BY for aggregations (sales, quantity, averages)

### Orders per Region

In [14]:
query = """
SELECT Region,COUNT(*) AS orders
FROM sales_data
GROUP BY Region
ORDER BY orders DESC
"""

pd.read_sql(query, conn)

,Region,orders
0,West,3203
1,East,2848
2,Central,2323
3,South,1620


The above  query shows the total orders in different regions.

### Sales per Region

In [15]:
query = """
SELECT Region,SUM(Sales) AS total_sales
FROM sales_data
GROUP BY Region
ORDER BY total_sales DESC
"""

pd.read_sql(query, conn)

,Region,total_sales
0,West,725457.8245
1,East,678781.2400
2,Central,501239.8908
3,South,391721.9050


The above query identifies which region generated the highest sales : West.

### Top 3 categories with highest avg_sales

In [21]:
query = """
SELECT Category, AVG(Sales) AS avg_sales
FROM sales_data
GROUP BY Category
ORDER BY avg_sales DESC
"""

pd.read_sql(query, conn)

,Category,avg_sales
0,Technology,452.709276
1,Furniture,349.834887
2,Office Supplies,119.324101


This query shows the average sales value for each product category.

### Top 3 categories with highest avg_sales

In [22]:
query = """
SELECT Category,
       SUM(Profit) AS total_profit
FROM sales_data
GROUP BY Category
ORDER BY total_profit DESC
"""

pd.read_sql(query, conn)

,Category,total_profit
0,Technology,145454.9481
1,Office Supplies,122490.8008
2,Furniture,18451.2728


This query identifies which product category generated the highest overall profit.

### MIN and MAX Profit

In [24]:
query = """
SELECT MIN(Sales) AS min_sales, MAX(Sales) AS max_sales
FROM sales_data
"""

pd.read_sql(query, conn)

,min_sales,max_sales
0,0.444,22638.48


This query shows the lowest and highest sales values.

### MIN and MAX Profit

In [25]:
query = """
SELECT MIN(Profit) AS min_profit, MAX(Profit) AS max_profit
FROM sales_data
"""

pd.read_sql(query, conn)

,min_profit,max_profit
0,-6599.978,8399.976


This query identifies the maximum profit earned and the maximum loss incurred in a single transaction.

## 5.Sort and limit results (top products, top categories)

### Top 10 Products by Sales

In [26]:
query = """
SELECT "Product Name", Sales , Profit
FROM sales_data
ORDER BY Sales DESC
LIMIT 10
"""

pd.read_sql(query, conn)

,Product Name,Sales,Profit
0,Cisco TelePresence System EX90 Videoconferenci...,22638.480,-1811.0784
1,Canon imageCLASS 2200 Advanced Copier,17499.950,8399.9760
2,Canon imageCLASS 2200 Advanced Copier,13999.960,6719.9808
3,Canon imageCLASS 2200 Advanced Copier,11199.968,3919.9888
4,Canon imageCLASS 2200 Advanced Copier,10499.970,5039.9856
5,GBC Ibimaster 500 Manual ProClick Binding System,9892.740,4946.3700
6,Ibico EPK-21 Electric Binding System,9449.950,4630.4755
7,"3D Systems Cube Printer, 2nd Generation, Magenta",9099.930,2365.9818
8,HP Designjet T520 Inkjet Large Format Printer ...,8749.950,2799.9840
9,Canon imageCLASS 2200 Advanced Copier,8399.976,1119.9968


This query identifies the products with the highest sales value in the dataset.

### Top Categories by Total Sales

In [27]:
query = """
SELECT Category , SUM(Sales) AS total_sales
FROM sales_data
GROUP BY Category
ORDER BY total_sales DESC
LIMIT 3
"""

pd.read_sql(query, conn)

,Category,total_sales
0,Technology,836154.0330
1,Furniture,741999.7953
2,Office Supplies,719047.0320


This query ranks product categories based on total sales and highlights the best-performing category.

## 6.Solve use cases (trends, top customers, duplicates)

### Top 10 Customers by Total Sales

In [28]:
query = """
SELECT "Customer Name" , SUM(Sales) AS total_sales
FROM sales_data
GROUP BY "Customer Name"
ORDER BY total_sales DESC
LIMIT 10
"""

pd.read_sql(query, conn)

,Customer Name,total_sales
0,Sean Miller,25043.050
1,Tamara Chand,19052.218
2,Raymond Buch,15117.339
3,Tom Ashbrook,14595.620
4,Adrian Barton,14473.571
5,Ken Lonsdale,14175.229
6,Sanjit Chand,14142.334
7,Hunter Lopez,12873.298
8,Sanjit Engle,12209.438
9,Christopher Conant,12129.072


This query identifies the customers who contributed the highest total sales.

### Most Profitable Customers

In [30]:
query = """
SELECT
    "Customer Name", SUM(Profit) AS total_profit
FROM sales_data
GROUP BY "Customer Name"
ORDER BY total_profit DESC
LIMIT 10
"""

pd.read_sql(query, conn)

,Customer Name,total_profit
0,Tamara Chand,8981.3239
1,Raymond Buch,6976.0959
2,Sanjit Chand,5757.4119
3,Hunter Lopez,5622.4292
4,Adrian Barton,5444.8055
5,Tom Ashbrook,4703.7883
6,Christopher Martinez,3899.8904
7,Keith Dawkins,3038.6254
8,Andy Reiter,2884.6208
9,Daniel Raglin,2869.0760


This query identifies the customers contributing the highest overall profit.

### Yearly Sales Trend

In [33]:
query = """
SELECT
    SUBSTR("Order Date",-4) AS year,
    SUM(Sales) AS total_sales
FROM sales_data
GROUP BY year
ORDER BY year
"""

pd.read_sql(query, conn)

,year,total_sales
0,2014,484247.4981
1,2015,470532.5090
2,2016,609205.5980
3,2017,733215.2552


This query shows how sales changed across different years, to identify seasonal trends.

### Duplicate Records Check

In [36]:
query = """
SELECT  "Row ID" ,  COUNT(*) AS occurs
FROM sales_data
GROUP BY "Row ID"
HAVING COUNT(*) > 1
"""

pd.read_sql(query, conn)

,Row ID,occurs


This query identifies any Row IDs appearing more than once in the dataset.


### Check Duplicate Order IDs

In [39]:
query = """
SELECT
    "Order ID",  COUNT(*) AS items
FROM sales_data
GROUP BY "Order ID"
HAVING COUNT(*) > 1
ORDER BY items DESC
LIMIT 10
"""

pd.read_sql(query, conn)

,Order ID,items
0,CA-2017-100111,14
1,CA-2017-157987,12
2,US-2016-108504,11
3,CA-2016-165330,11
4,US-2015-126977,10
5,CA-2016-105732,10
6,CA-2015-131338,10
7,US-2016-114013,9
8,US-2015-163433,9
9,CA-2017-140949,9


This query identifies orders that contain multiple items. Multiple occurrences of the same Order ID are expected because a single order can include more than one product.

### CASE on Profit

In [41]:
query = """
SELECT
CASE
    WHEN Profit > 0 THEN 'Profit'
    ELSE 'Loss'
END AS order_status,
COUNT(*) AS total_orders
FROM sales_data
GROUP BY order_status
"""

pd.read_sql(query, conn)

,order_status,total_orders
0,Loss,1936
1,Profit,8058


This query classifies orders into profitable and loss-making orders and shows their distribution.

### Compare Sales using different Shipping Modes

In [45]:
query = """
SELECT
    "Ship Mode",
    COUNT(*) AS total_orders,
    ROUND(SUM(Sales),2) AS total_sales
FROM sales_data
GROUP BY "Ship Mode"
ORDER BY total_sales DESC
"""

pd.read_sql(query, conn)

,Ship Mode,total_orders,total_sales
0,Standard Class,5968,1358215.74
1,Second Class,1945,459193.57
2,First Class,1538,351428.42
3,Same Day,543,128363.13


This query compares different shipping modes based on the number of orders and total sales generated.

## 7.Validate results (row counts, data quality)

### Verify Total Records

In [46]:
query = """
SELECT COUNT(*) AS total_rows
FROM sales_data
"""

pd.read_sql(query, conn)

,total_rows
0,9994


This query verifies the total number of records loaded into the database.

### Check for Missing Sales Values

In [47]:
query = """
SELECT COUNT(*) AS missing_sales
FROM sales_data
WHERE Sales IS NULL
"""

pd.read_sql(query, conn)

,missing_sales
0,0


This query checks whether any records have missing sales values.

### Check for Missing Customer IDs

In [49]:
query = """
SELECT COUNT(*) AS missing_cust_ids
FROM sales_data
WHERE "Customer ID" IS NULL
"""
pd.read_sql(query, conn)

,missing_cust_ids
0,0


This query checks for missing customer identifiers in the dataset.